# INFO 370 2021 Wi - Python Pandas Basics

*Name:* Peachyapa Saengcharoentrakul

In [ ]:
### Instructions

1. Please complete the lab tutorial during your respective lab session.

2. Please write down your name and your collaborators (if any).

3. In this exercise you will get hands-on experience in importing data into structured format, summarizing data using descriptive statistics (e,g, sum, average, etc.) and manipulating data including indexing, slicing and grouping.

4. The data you will be working with is the pulled from Johns Hopkins University COVID-19 Data Repository, and we will be focusing on confirmed COVID-19 deaths throughought the US. It is a daily record, collected from local and state health departments.  You can find [additional information](https://github.com/CSSEGISandData/COVID-19/blob/master/csse_covid_19_data/README.md) and [updated data](https://github.com/CSSEGISandData/COVID-19/tree/master/csse_covid_19_data/csse_covid_19_time_series) on github.

Some of the more unclear variables:
    * FIPS: US only. Federal Information Processing Standards code that uniquely identifies counties within the USA.
    * Admin2: County name
    * UID: Similar to Admin2, but includes locations like Diamond Princess, unassigned, etc. (Not needed for this activity)
    * "6/16/2020": total number of confirmed COVID-19 deaths as of June 16, 2020

5. For each question, please type your codesin a *Code* cell and your written response which summarizes your results in a *Markdown* cell.

6. Don't be scared. We are here to help you learn. :)

In [15]:
# import packages
import numpy as np
import pandas as pd
import os
# you may import other packages if necessary

### 1. Data Import and Data Summary

**(a) Load the data into a pandas dataframe.**

Use `pd.read_csv` and check what is the right separator (`sep=...`).  Print the first few lines of it as a sanity check.

Note: `pd.read_csv` can read compressed files directly, you do not have to decompress those.

In [33]:
# Your codes here
data = pd.read_csv("data/time_series_covid19_deaths_US-reduced.csv", sep="\t")

**(b) It's time to get to know your data! Report the number of rows and columns in the dataset.**

In [17]:
data.shape

(3340, 29)

Number of columns: 29
Number of rows: 3340

**(c) What variables does this dataset have? Report the variable names along with the data type of each variable.**

In [18]:
# Your codes here
data.dtypes

UID                 int64
iso2               object
FIPS              float64
Admin2             object
Province_State     object
Lat               float64
Long_             float64
Combined_Key       object
Population          int64
2/1/20              int64
3/1/20              int64
4/1/20              int64
5/1/20              int64
6/1/20              int64
7/1/20              int64
8/1/20              int64
9/1/20              int64
10/1/20             int64
11/1/20             int64
12/1/20             int64
1/1/21              int64
1/2/21              int64
1/3/21              int64
1/4/21              int64
1/5/21              int64
1/6/21              int64
1/7/21              int64
1/8/21              int64
1/9/21              int64
dtype: object

**(d) What is the number of NULL/NA values in each column of the dataframe?**

In [19]:
# Your codes here
data.isnull().sum(axis=0) #I am googling half of this https://stackoverflow.com/questions/22257527/how-do-i-get-a-summary-count-of-missing-nan-data-by-column-in-pandas
#alternatively, the total NAs data.isnull().sum().sum()

UID                0
iso2               0
FIPS              10
Admin2             6
Province_State     0
Lat                0
Long_              0
Combined_Key       0
Population         0
2/1/20             0
3/1/20             0
4/1/20             0
5/1/20             0
6/1/20             0
7/1/20             0
8/1/20             0
9/1/20             0
10/1/20            0
11/1/20            0
12/1/20            0
1/1/21             0
1/2/21             0
1/3/21             0
1/4/21             0
1/5/21             0
1/6/21             0
1/7/21             0
1/8/21             0
1/9/21             0
dtype: int64

### 2. Data Manipulation - Explore Death Tolls by State

**(a) What are the 'states/provices' with the most confirmed deaths as of 6/16/2020?**

Hint: check out _DataFrame.groupby_ method, _Series.sum()_ method, and _Series.sort__values()_ method.

In [72]:
by_state = data.groupby(['Province_State'])['6/1/20'].sum().sort_values(ascending=False)

What is the data structure you got?  What is its index and what is its value?

Hint: use `type` for data type, and `.index` and `.values` for the latter.

In [73]:
type(by_state)
by_state.index

Index(['New York', 'New Jersey', 'Massachusetts', 'Michigan', 'Pennsylvania',
       'Illinois', 'California', 'Connecticut', 'Louisiana', 'Maryland',
       'Florida', 'Ohio', 'Indiana', 'Georgia', 'Texas', 'Colorado',
       'Virginia', 'Washington', 'Minnesota', 'North Carolina', 'Arizona',
       'Rhode Island', 'Missouri', 'Mississippi', 'Wisconsin', 'Alabama',
       'Iowa', 'South Carolina', 'District of Columbia', 'Kentucky', 'Nevada',
       'Delaware', 'Tennessee', 'New Mexico', 'Oklahoma', 'New Hampshire',
       'Kansas', 'Nebraska', 'Oregon', 'Puerto Rico', 'Arkansas', 'Utah',
       'Maine', 'Idaho', 'West Virginia', 'South Dakota', 'North Dakota',
       'Vermont', 'Hawaii', 'Montana', 'Wyoming', 'Alaska', 'Virgin Islands',
       'Guam', 'Grand Princess', 'Northern Mariana Islands',
       'Diamond Princess', 'American Samoa'],
      dtype='object', name='Province_State')

In [74]:
by_state.values

array([30832, 13547,  7035,  5758,  5567,  5412,  4217,  3970,  2801,
        2552,  2460,  2207,  2143,  2094,  1879,  1458,  1392,  1128,
        1060,   948,   918,   825,   815,   739,   656,   646,   555,
         500,   468,   439,   417,   368,   364,   362,   334,   245,
         217,   170,   154,   136,   133,   113,    89,    82,    76,
          62,    61,    55,    17,    17,    17,    10,     6,     5,
           3,     2,     0,     0])

**(b) The 'states/provinces' column contains more than just the 50 States and DC. Please filter your results from 2a but only include the 50 states and DC**

Below is the list of states made for you.

Hint: use `pd.Series.isin()` method.

In [22]:
states_list = ['Alabama', 'Alaska', 'Arizona', 'Arkansas', 'California', 'Colorado', 'Connecticut', 'Delaware', 'District of Columbia', 'Florida', 'Georgia', 'Hawaii', 'Idaho', 'Illinois', 'Indiana', 'Iowa', 'Kansas', 'Kentucky', 'Louisiana', 'Maine', 'Maryland', 'Massachusetts', 'Michigan', 'Minnesota', 'Mississippi', 'Missouri', 'Montana', 'Nebraska', 'Nevada', 'New Hampshire', 'New Jersey', 'New Mexico', 'New York', 'North Carolina', 'North Dakota', 'Ohio', 'Oklahoma', 'Oregon', 'Pennsylvania', 'Rhode Island', 'South Carolina', 'South Dakota', 'Tennessee', 'Texas', 'Utah', 'Vermont', 'Virginia', 'Washington', 'West Virginia', 'Wisconsin', 'Wyoming']

In [23]:
# answer here...
by_state[by_state.index.isin(states_list)]

Province_State
New York                30832
New Jersey              13547
Massachusetts            7035
Michigan                 5758
Pennsylvania             5567
Illinois                 5412
California               4217
Connecticut              3970
Louisiana                2801
Maryland                 2552
Florida                  2460
Ohio                     2207
Indiana                  2143
Georgia                  2094
Texas                    1879
Colorado                 1458
Virginia                 1392
Washington               1128
Minnesota                1060
North Carolina            948
Arizona                   918
Rhode Island              825
Missouri                  815
Mississippi               739
Wisconsin                 656
Alabama                   646
Iowa                      555
South Carolina            500
District of Columbia      468
Kentucky                  439
Nevada                    417
Delaware                  368
Tennessee                

**(c) What is the total number of confirmed deaths in Washington State as of 1/9/2021?**

In [105]:
total_WA = data.groupby(['Province_State'])['1/9/21'].sum().sort_values(ascending=False) #1/9/2021 instead of 2020
total_WA

Province_State
New York                    39471
Texas                       30313
California                  29707
Florida                     22804
New Jersey                  19854
Illinois                    19210
Pennsylvania                17626
Michigan                    14145
Massachusetts               13074
Georgia                     11457
Arizona                     10036
Ohio                         9599
Indiana                      8966
Louisiana                    7833
Tennessee                    7704
North Carolina               7425
Connecticut                  6324
Maryland                     6246
Missouri                     6129
South Carolina               5758
Minnesota                    5731
Wisconsin                    5567
Virginia                     5381
Alabama                      5299
Colorado                     5190
Mississippi                  5146
Iowa                         4127
Arkansas                     4010
Washington                   3698

In [120]:
total_WA.take([28])

Province_State
Washington    3698
Name: 1/9/21, dtype: int64

In Washington State, there are 3698 confirmed deaths as of 1/9/21

**(d) What is the average number of deaths for a county in Washington State, how does it compare to average number of cases  for a county nationally?**

In [70]:
#for a particular date?
#average number of deahts in WA state
WA_average = data[data.Province_State == "Washington"].mean()
WA_avg = WA_average["1/9/21"] #on Jan 9 '21' the average # deaths across different counties in WA is 90
WA_avg

90.1951219512195

In [71]:
national_average = data.groupby(['Province_State'])['1/9/21'].mean().sort_values(ascending=False) #get the mean for each state
national_average = national_average.mean() #then get the mean for the national
national_average

141.61851898942862

National average = 141 people (rounded down)
Washington average number of deaths per county = 90 people (rounded down)

### 3. Extra task (not graded)

If you still have time/interest, then try to understand the issues in data.

**(a) In 1c, we found that some columns have NULL/NA values. Briefly look at those values. Can you understand why those rows are included as separate entries? Googling some entries may help, as well as referring to the Github data source for variable definitions. Furthermore, how does this hypothesis affect how you would go about analyzing the dataset?**

In [27]:
# look at data

In [14]:
# look at data